### PR TITLE
feat: bundle dependencies

### LINKED ISSUE #338: Bundling pnpm dependencies
The idea is to make pnpm installation super fast. If pnpm installation is fast enough, we can recommend to use it instead of npm on travis/appveyor installations. pnpm installation won't increase significantly the setup time but will decrease significantly the installation time of the dependencies.

To make pnpm installation fast, all its dependencies have to be bundled with the package.

Possible solutions:

1. using the `bundledDependencies` feature of npm. See branch [bundled-deps-2](https://github.com/rstacruz/pnpm/tree/bundled-deps-2)
2. using some hacks that will allow to publish the content of the `node_modules` with pnpm. Working implementation: https://github.com/rstacruz/pnpm/pull/310
3. using a bundler like rollup.

TODO: add Pros and Cons of each solution

### PR DESCRIPTION
**Pull request by zkochan**
Close #338

@andreypopp @iamstarkov @rstacruz. This is unbelievable! Seems like npm has a secret way to publish `node_modules` together with the package. I discovered it in the bower package. If the `node_modules` folder is in the `lib/` folder, it will be published with the package!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -106,10 +106,10 @@
     "lint": "standard && tslint -c tslint.json ./src/**/*.ts ./test/**/*.ts",
     "test": "npm run lint && ts-node test | tap-diff",
     "posttest": "rm -rf .tmp",
-    "prerelease": "npm test",
+    "prerelease": "npm run tsc && npm test",
     "release": "sh .scripts/release.sh",
     "tsc": "rm -rf lib && tsc",
-    "prepublish": "npm run tsc"
+    "prepublish": "in-publish && echo 'You need to use \"npm run release\" to publish pnpm' && false || not-in-publish"
```
Comment (zkochan) [inline package.json:111]: Or maybe I can put the code in here from release.sh. I'll try.

Comment (zkochan) [reply package.json:111]: Seems like I can't modify the package.json from a prepublish script, so this implementation is the only way

File: package.json [hunk 2]
```diff
@@ -13,9 +13,7 @@
   "typings": "lib/index.d.ts",
   "files": [
     "bin/",
-    "lib/",
-    ".scripts/",
-    "cached_node_modules/"
+    "lib/"
```
Comment (iamstarkov) [inline package.json:16]: if node_modules are in `.tmp/`, not in `lib/`, then this line might be incorrect

**Comment by iamstarkov** [comment]
thats unbelievable indeed

**Comment by iamstarkov** [comment]
though npm explicitly says it blocks [root node_modules](https://docs.npmjs.com/misc/developers#keeping-files-out-of-your-package), so i think any other node_modules are fine

File: package.json (hunk 2 above)
Comment (zkochan) [reply package.json:16]: it is in `lib/`. I move the current node_modules temporarily to `.tmp/` and install a fresh bundle. After publishing the initial node_modules is moved back to the root

**Comment by andreypopp** [comment]
@zkochan this is awesome! Going to bother this approach for my own Node CLI tools. pnpm now installs super fast.

**Comment by zkochan** [comment]
Here are the improvements I see in Travis, when running `npm -g pnpm`:

**using npm2**

* before: `31.38s`
* after: `8.76s`

**using npm3:**

* before: `15.37s`
* after: `8.42s`

sometimes it is even less than 8 seconds

**Comment by zkochan** [comment]
To be honest, it still feels too slow to me:smile:

I think we can make it even faster if we won't use npm for installation, but that's a smaller priority for now than getting to v1

**Pull request merged**